### PR TITLE
Plugins: Add sql support for the secure socks proxy

### PR DESCRIPTION
**Pull request by stephaniehingtgen**
**What is this feature?**

This PR is a follow up to https://github.com/grafana/grafana/pull/59254, to add SQL datasources to the supported datasources for the feature toggle, `secureSocksDatasourceProxy`. This feature allows supported datasources to go through a socks5 proxy with TLS if the data source has `enableSecureSocksProxy=true` in the json data.

**Why do we need this feature?**

Grafana users need to be able to connect to datasources that live in different networks than where Grafana is running.

**Who is this feature for?**

Grafana users who are running datasources in multiple networks and cannot open ports up to the public internet.

**Which issue(s) does this PR fix?**:

Directly relates to: https://github.com/grafana/hosted-grafana/issues/2922
Also related to: https://github.com/grafana/grafana/issues/15045

**Screenshots of the UI & the connections working**:
mssql
postgres
mysql

### CODE REVIEW

File: pkg/infra/proxy/secure_socks_proxy.go [hunk 1]
```diff
@@ -14,24 +14,40 @@ import (
 	"golang.org/x/net/proxy"
 )
 
-// newSecureSocksProxy takes a http.DefaultTransport and wraps it in a socks5 proxy with TLS
-func newSecureSocksProxy(cfg *setting.SecureSocksDSProxySettings, transport *http.Transport) error {
+// NewSecureSocksProxy takes a http.DefaultTransport and wraps it in a socks5 proxy with TLS
```
Comment (xnyo) [inline pkg/infra/proxy/secure_socks_proxy.go:17]:
```suggestion
// NewSecureSocksHTTPProxy takes a http.DefaultTransport and wraps it in a socks5 proxy with TLS
```

**Review by xnyo — COMMENTED** [review]
Great work! Some comments:

File: pkg/infra/proxy/secure_socks_proxy.go (hunk 1 above)
Comment (stephaniehingtgen) [reply pkg/infra/proxy/secure_socks_proxy.go]: Good catch - updated!

File: pkg/tsdb/mssql/proxy.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+func createMSSQLProxyDriver(settings *setting.SecureSocksDSProxySettings, cnnstr string) (string, error) {
+	// create a unique driver per connection string
+	hash, err := util.Md5SumString(cnnstr)
+	if err != nil {
+		return "", err
+	}
+	driverName := "mssql-proxy-" + hash
+
+	// only register the driver once
+	if core.QueryDriver(driverName) == nil {
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply pkg/tsdb/mssql/proxy.go:32]: Oh dang, really good catch there! I updated all the xorm driver queries and registers to be behind a RW mutex now

**Review by marefr — COMMENTED** [review]
Added some comments. Agree with earlier review. It's unfortunate we have to do all this to enable this - there's no other way to override the dialer? :cry:

**Comment by stephaniehingtgen** [comment]
> It's unfortunate we have to do all this to enable this - there's no other way to override the dialer? 😢

I agree :( I spent a fair amount of time over the past few weeks trying to find a better way to do it, but this was the cleanest solution I could find.

For mssql: You can [update the connector to update the dialer](https://github.com/denisenkom/go-mssqldb/blob/e53873151a46af22b77fa010fc1485c31cbb38b1/mssql.go#L142). However, this can only be used through [sql.OpenDB](https://pkg.go.dev/database/sql#OpenDB), but [xorm uses](https://gitea.com/xorm/core/src/branch/master/db.go#L86) just regular [open](https://pkg.go.dev/database/sql#Open). So there is no way to override the connector. 

For postgres, it's the same thing, the connector is updated [here](https://github.com/lib/pq/blob/master/connector.go#L31), which isn't used in xorm. 

MySQL is nice because they allow you to update the dialer by saving it as a "network" [here](https://github.com/go-sql-driver/mysql/blob/0c62bb2791485d4260371bcc6017321de93b2430/driver.go#L49), that can then be used in the connection string.

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -81,6 +81,7 @@
 /pkg/infra/metrics/ @grafana/backend-platform
 /pkg/infra/network/ @grafana/backend-platform
 /pkg/infra/process/ @grafana/backend-platform
+/pkg/infra/proxy/ @grafana/hosted-grafana-team @grafana/backend-platform
```
Comment (tolzhabayev) [inline .github/CODEOWNERS:84]: 1. CODEOWNERS file is out of sync with main
2. I don't think backend-platform has anything to do with this as this is a new package, so I would suggest to keep it to hosted grafana

**Review by tolzhabayev — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -81,6 +81,7 @@
 /pkg/infra/metrics/ @grafana/backend-platform
 /pkg/infra/network/ @grafana/backend-platform
 /pkg/infra/process/ @grafana/backend-platform
+/pkg/infra/proxy/ @grafana/backend-platform
```
[the remark this answers is not in the document]
Comment (stephaniehingtgen) [reply .github/CODEOWNERS]: Updated to just be the hg team 👍

File: .github/CODEOWNERS (hunk 3 above)
Comment (stephaniehingtgen) [reply .github/CODEOWNERS]: 👍 Merged in main and updated to be just the HG team

**Review by tolzhabayev — APPROVED** [review]

**Review by marefr — APPROVED** [review]
LGTM

**Review by xnyo — APPROVED** [review]
LGTM!

**Pull request merged**